sklearn/mixture/gaussian_mixture.py

```diff
@@ -336,110 +336,99 @@ def _compute_precision_cholesky(covariances, covariance_type):
-    log_det = -2. * np.sum(np.log(np.diagonal(precision_chol)))
-    for k, mu in enumerate(means):
-        y = np.dot(X - mu, precision_chol)
-        log_prob[:, k] = np.sum(np.square(y), axis=1)
-    log_prob = -.5 * (n_features * np.log(2. * np.pi) + log_det + log_prob)
-    return log_prob
 
-
-def _estimate_log_gaussian_prob_diag(X, means, precisions_chol):
-    """Estimate the log Gaussian probability for 'diag' precision.
+def _estimate_log_gaussian_prob(X, means, precisions_chol, covariance_type):
+    """Estimate the log Gaussian probability.
 
     Parameters
     ----------
     X : array-like, shape (n_samples, n_features)
 
     means : array-like, shape (n_components, n_features)
 
-    precisions_chol : array-like, shape (n_components, n_features)
+    precisions_chol : array-like,
         Cholesky decompositions of the precision matrices.
+        'full' : shape of (n_components, n_features, n_features)
+        'tied' : shape of (n_features, n_features)
+        'diag' : shape of (n_components, n_features)
+        'spherical' : shape of (n_components,)
 
-    Returns
-    -------
-    log_prob : array-like, shape (n_samples, n_components)
-    """
-    n_samples, n_features = X.shape
-    precisions = precisions_chol ** 2
-    log_prob = -.5 * (n_features * np.log(2. * np.pi) -
-                      np.sum(np.log(precisions), 1) +
-                      np.sum((means ** 2 * precisions), 1) -
-                      2. * np.dot(X, (means * precisions).T) +
-                      np.dot(X ** 2, precisions.T))
-    return log_prob
-
-
-def _estimate_log_gaussian_prob_spherical(X, means, precisions_chol):
-    """Estimate the log Gaussian probability for 'spherical' precision.
-
-    Parameters
-    ----------
-    X : array-like, shape (n_samples, n_features)
-
-    means : array-like, shape (n_components, n_features)
-
-    precisions_chol : array-like, shape (n_components, )
-        Cholesky decompositions of the precision matrices.
+    covariance_type : {'full', 'tied', 'diag', 'spherical'}
 
     Returns
     -------
-    log_prob : array-like, shape (n_samples, n_components)
+    log_prob : array, shape (n_samples, n_components)
     """
     n_samples, n_features = X.shape
-    precisions = precisions_chol ** 2
-    log_prob = -.5 * (n_features * np.log(2 * np.pi) -
-                      n_features * np.log(precisions) +
-                      np.sum(means ** 2, 1) * precisions -
-                      2 * np.dot(X, means.T * precisions) +
-                      np.outer(np.sum(X ** 2, axis=1), precisions))
-    return log_prob
+    n_components, _ = means.shape
+    # det(precision_chol) is half of det(precision)
+    log_det = _compute_log_det_cholesky(
+        precisions_chol, covariance_type, n_features)
+
+    if covariance_type == 'full':
+        log_prob = np.empty((n_samples, n_components))
+        for k, (mu, prec_chol) in enumerate(zip(means, precisions_chol)):
+            y = np.dot(X - mu, prec_chol)
```

Comment:
y = np.dot(X - mu, prec_chol)
->
y = np.dot(X, prec_chol) - mu[:, np.newaxis] * prec_chol

would avoid a copy of X